pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,267 @@
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	vdbcontroller "github.com/vertica/vertica-kubernetes/pkg/controllers/vdb"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/unsandboxsc"
+	"github.com/vertica/vertica-kubernetes/pkg/vdbstatus"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UnsandboxSubclusterReconciler will unsandbox subclusters in the sandbox of a sandbox config map
+type UnsandboxSubclusterReconciler struct {
+	SRec *SandboxConfigMapReconciler
+	Vdb  *vapi.VerticaDB
+	Log  logr.Logger
+	client.Client
+	Dispatcher  vadmin.Dispatcher
+	PFacts      *vdbcontroller.PodFacts
+	ConfigMap   *corev1.ConfigMap
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+}
+
+func MakeUnsandboxSubclusterReconciler(r *SandboxConfigMapReconciler, vdb *vapi.VerticaDB, log logr.Logger,
+	cli client.Client, pfacts *vdbcontroller.PodFacts, dispatcher vadmin.Dispatcher, configMap *corev1.ConfigMap) controllers.ReconcileActor {
+	pfactsForMainCluster := pfacts.Copy(vapi.MainCluster)
+	return &UnsandboxSubclusterReconciler{
+		SRec:       r,
+		Log:        log.WithName("SandboxControllerUnsandboxSubclusterReconciler"),
+		Vdb:        vdb,
+		Client:     cli,
+		Dispatcher: dispatcher,
+		PFacts:     &pfactsForMainCluster,
+		ConfigMap:  configMap,
+	}
+}
+
+// Reconcile will unsandbox subclusters, remove the sandbox config map, and update sandbox status of vdb
+func (r *UnsandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db
+	if r.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly || !r.Vdb.IsEON() {
+		return ctrl.Result{}, nil
+	}
+
+	// collect pod facts for main cluster
+	if err := r.PFacts.Collect(ctx, r.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// reconcile sandbox status for the subclusters that are already unsandboxed
+	if err := r.reconcileSandboxStatus(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// reconcile the sandbox config map if it expires
+	err, deleted := r.reconcileSandboxConfigMap(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// if we deleted the expired config map, we don't need to make unsandbox operation any more
+	if deleted {
+		return ctrl.Result{}, nil
+	}
+
+	// only execute unsandbox op when unsandbox trigger id and sandbox name are set
+	if r.ConfigMap.Annotations[vmeta.SandboxControllerUnsandboxTriggerID] != "" &&
+		r.ConfigMap.Data[vapi.SandboxNameKey] != "" {
+		// we ignore the config map that contains wrong data,
+		// this kind of config map could be old or created by others
+		if r.ConfigMap.Data[vapi.VerticaDBNameKey] != r.Vdb.Name {
```

Comment:
Is this a legit error? The sandbox controller will fetch the vdb based on what's in the ConfigMap. So, this invariant should always hold.

Reply:
You are correct. Will remove this check.